src/choose_gossip_peer_strategy.rs

```diff
@@ -0,0 +1,306 @@
+use crdt::ReplicatedData;
+use rand::distributions::{IndependentSample, Weighted, WeightedChoice};
+use rand::thread_rng;
+use result::{Error, Result};
+use signature::PublicKey;
+use std;
+use std::collections::HashMap;
+
+pub const DEFAULT_WEIGHT: u32 = 1;
+
+pub trait ChooseGossipPeerStrategy {
+    fn choose_peer(&self, options: Vec<&ReplicatedData>) -> Result<ReplicatedData>;
+}
+
+pub struct ChooseRandomPeerStrategy<'a> {
+    random: &'a Fn() -> u64,
+}
+
+impl<'a> ChooseRandomPeerStrategy<'a> {
+    pub fn new(random: &'a Fn() -> u64) -> Self {
+        ChooseRandomPeerStrategy { random }
+    }
+}
+
+impl<'a> ChooseGossipPeerStrategy for ChooseRandomPeerStrategy<'a> {
+    fn choose_peer(&self, options: Vec<&ReplicatedData>) -> Result<ReplicatedData> {
+        if options.len() < 1 {
+            return Err(Error::CrdtTooSmall);
+        }
+
+        let n = ((self.random)() as usize) % options.len();
+        Ok(options[n].clone())
+    }
+}
+
+pub struct ChooseWeightedPeerStrategy<'a> {
+    remote: &'a HashMap<PublicKey, u64>,
+    external_liveness: &'a HashMap<PublicKey, HashMap<PublicKey, u64>>,
```

Comment:
Can you comment each field? Especially `external_liveness`. Is this name precise? What's the difference between `remote` and `external`? I think I'm missing something about "liveness" too. Is the node's liveness relevant here? Maybe that's a question for @aeyakovenko

Reply:
That's a fair point, @aeyakovenko suggested it so I think he might have more insight. Perhaps remote_rumors might be a better name